src/shiki-bridge-theme.ts

```diff
@@ -0,0 +1,103 @@
+import type { ThemeRegistration, ThemeRegistrationRaw } from "shiki";
+import type { ExtensionTheme } from "vscode-extension-manifest";
+import type { ThemeRegistry } from "./theme-registry.js";
+import { type ExtensionFileReader } from "./vscode-utils.js";
+import { logger } from "./logger.js";
+
+type ThemeRegistrationRawWithInclude = ThemeRegistrationRaw & { include?: string };
+
+export async function buildThemeRegistration(contribution: ExtensionTheme, registry: ThemeRegistry, fileReader: ExtensionFileReader, Uri: typeof import('vscode').Uri): Promise<ThemeRegistration> {
+    const id = contribution.id ?? contribution.label;
+
+    if (!id) {
+        logger.trace(`cannot build a theme registration for a contribution without a label or id`, contribution);
+        return {};
+    }
+
+    /**
+     * Because the `include` property is relative to the file it is in, we need to keep track of the location of the last included grammar.
+     */
+    const rawTheme = await resolveThemeGrammar(contribution, registry, fileReader, Uri);
+
+    if (typeof rawTheme.tokenColors === 'string') {
+        logger.trace(`theme '${id}' has 'tokenColors' set to a relative file path`, contribution, rawTheme);
+    }
+
+    const themeRegistration: ThemeRegistration = {
+        name: id,
+        displayName: contribution.label,
+        /**
+         * NOTE: note that we use `settings` instead of `tokenColors` when building the final theme registration, as this is what Shiki expects
+         */
+        settings: rawTheme.tokenColors,
+        colors: rawTheme.colors,
+        semanticHighlighting: rawTheme.semanticHighlighting,
+        semanticTokenColors: rawTheme.semanticTokenColors,
+        type: bridgeThemeType(contribution.uiTheme),
+    };
+
+    return themeRegistration;
+}
+
+type ThemeMergeResult = Pick<ThemeRegistrationRawWithInclude, 'colors' | 'include' | 'semanticHighlighting' | 'semanticTokenColors' | 'type' | 'tokenColors'>;
+
+/**
+ * Recursively resolves the `include` property of theme grammars.
+ *
+ * NOTE: the `tokenColors` can be a string which would make it a relative file path to another TextMate file
+ *       this does not seem to be used in any theme, but if it does happen we log it and set it to an empty array as Shiki expects it to be an array
+ */
+export async function resolveThemeGrammar(contribution: ExtensionTheme, registry: ThemeRegistry, fileReader: ExtensionFileReader, Uri: typeof import('vscode').Uri) {
```

Comment:
As far as I understand we can use `Theme.createFromRawTheme` from "vscode-textmate" to recuresively resolve the include statements and merge it

Reply:
Unfortunately, `vscode-textmate` uses only `IRawTheme` objects, which are already resolved.
The `vscode` theme loading is done here: https://github.com/microsoft/vscode/blob/main/src/vs/workbench/services/themes/common/colorThemeData.ts#L712. And as all vscode source code its not really extractable

Reply:
Oh, what a bummer :/